File: COSC541 Foundations of Automata and Languages/All Course Files/JAVA/TreeExamples.java
package automata.treedemo;

import java.util.Arrays;
import java.util.List;


public class TreeExamples {
	
    public static class Node implements DiagonalTree.TreeNode {
        private final String label;
        private final List<Node> children;

        public Node(String label, Node... children) {
            this.label = label;
            this.children = Arrays.asList(children);
        }

        @Override
        public String getLabel() {
            return label;
        }

        @Override
        public List<? extends DiagonalTree.TreeNode> getChildren() {
            return children;
        }
    }

    public static void main(String[] args) {
        // e1: (a+b)*ab
        Node tree_e1 = new Node("·",
            new Node("*",
                new Node("+", new Node("a"), new Node("b"))
            ),
            new Node("·", new Node("a"), new Node("b"))
        );

        // e2: (a + b)*
        Node tree_e2 = new Node("*",
            new Node("+", new Node("a"), new Node("b"))
        );

        // e3: (a + b)*abb
        Node tree_e3 = new Node("·",
            new Node("*",
                new Node("+", new Node("a"), new Node("b"))
            ),
            new Node("·",
                new Node("a"),
                new Node("·",
                    new Node("b"),
                    new Node("b")
                )
            )
        );

        // e3_2: (a + b)*abb
        Node tree_e3_2 = new Node("·",
            new Node("·",
                new Node("·",
                    new Node("*", new Node("+", new Node("a"), new Node("b"))),
                    new Node("a")
                ),
                new Node("b")
            ),
            new Node("b")
        );

        // e3_3: (a + b)*abb
        Node tree_e3_3 = new Node("·",
            new Node("·",
                new Node("*", new Node("+", new Node("a"), new Node("b"))),
                new Node("a")
            ),
            new Node("·",
                new Node("b"),
                new Node("b")
            )
        );

        // print tree
        DiagonalTree.printTree(tree_e1);
    }

}

File: COSC541 Foundations of Automata and Languages/All Course Files/PYTHON/tree_examples.py
from Common.diagonal_tree import print_diagonal_tree


# Tree node class
class Node:
    def __init__(self, label, children=None):
        self.label = label
        self.children = children or []


def printregex(node):
    if not node.children:
        print(node.label, end="")
        return
    elif not node.children[0]:
        print(node.label, end="")
    elif not node.children[1]:
        print(node.label, end="")
        return
    else:
        printregex(node.children[0])
        node.children[0] = []
        printregex(node)
        printregex(node.children[1])
        node.children[1] = []


# e1: (a+b)*ab
tree_e1 = Node("·", [
    Node("*", [
        Node("+", [Node("a"), Node("b")])
    ]),
    Node("·", [Node("a"), Node("b")])
])

# e2: (a + b)*
tree_e2 = Node("*", [
    Node("+", [Node("a"), Node("b")])
])

# e3: (a + b)*abb
tree_e3 = Node("·", [
    Node("*", [
        Node("+", [Node("a"), Node("b")])
    ]),
    Node("·", [
        Node("a"),
        Node("·", [
            Node("b"),
            Node("b")
        ])
    ])
])

tree_e3_2 = Node("·", [
    Node("·", [
        Node("·", [
            Node("*", [Node("+", [Node("a"), Node("b")])]),
            Node("a")
        ]),
        Node("b")
    ]),
    Node("b")
])

tree_e3_3 = Node("·", [
    Node("·", [
        Node("*", [Node("+", [Node("a"), Node("b")])]),
        Node("a")
    ]),
    Node("·", [
        Node("b"),
        Node("b")
    ])
])

tree_e4_1 = Node(".", [
    Node("*", [
        Node("+", [Node("a"), Node("b")])

    ]), Node("c")
])

tree_e4_2 = Node(".", [
    Node("+", [Node("a"), Node("b")]),
    Node("+", [Node("a"), Node("b")])
])

tree_e4_3 = Node(".", [
    Node("a"),
    Node("*", [
        Node("+", [
            Node("b"), Node("c")
        ])]
         )])

# print the tree
print_diagonal_tree(tree_e4_1)
print("----------------")
print_diagonal_tree(tree_e4_2)
print("----------------")
print_diagonal_tree(tree_e4_3)
printregex(tree_e4_2)

